django/utils/html.py

```diff
@@ -283,7 +283,7 @@ def trim_punctuation(lead, middle, trail):
             middle_unescaped = html.unescape(middle)
             stripped = middle_unescaped.rstrip(TRAILING_PUNCTUATION_CHARS)
             if middle_unescaped != stripped:
-                trail = middle[len(stripped):] + trail
+                trail = middle_unescaped[len(stripped):] + trail
```

Comment:
This version works fine, however it's a bit confusing that we concatenate escaped and unescaped string. I would use `middle` for both, e.g.
```python
punctuation_count = len(middle_unescaped) - len(stripped)
trail = middle[-punctuation_count:] + trail
middle = middle[:-punctuation_count]
```